lib/course-db.js

```diff
@@ -188,42 +302,53 @@ module.exports.checkInfoValid = function(idName, info, infoFile, courseInfo, log
 
 module.exports.loadInfoDB = function(db, idName, parentDir, infoFilename, defaultInfo, schema, optionSchemaPrefix, courseInfo, logger, callback) {
     var that = this;
+    // `cache` is an object with which we can cache information derived from course info
+    // in between successive calls to `checkInfoValid`
+    const cache = {};
+    beginSection(parentDir);
     fs.readdir(parentDir, function(err, files) {
         if (ERR(err, callback)) return;
 
-        async.filter(files, function(dirName, callback) {
-            // Filter out files from parentDir as it is possible they slip in without the user putting them there (like .DS_Store).
-            var filePath = path.join(parentDir, dirName);
-            fs.lstat(filePath, function(err, fileStats) {
+        // console.log(`beginning filter`, new Date());
+        async.each(files, function(dir, callback) {
+            var infoFile = path.join(parentDir, dir, infoFilename);
+            // console.log(`starting read of ${infoFile}`, new Date());
+            jsonLoad.readInfoJSON(infoFile, schema, function(err, info) {
+                if (err && err.code && err.path && (err.code === 'ENOTDIR') && err.path === infoFile) {
+                    // In a previous version of this code, we'd pre-filter
+                    // all files in the parent directory to remove anything
+                    // that may have accidentally slipped in, like .DS_Store.
+                    // However, that resulted in a huge number of system calls
+                    // that got really slow for large directories. Now, we'll
+                    // just blindly try to read a file from the directory and assume
+                    // that if we see ENOTDIR, that means the directory was not
+                    // in fact a directory.
```

Comment:
I don't think it's so much that the failed open was faster than an `lstat()` as it was that in practice we were doing a lot of unnecessary `lstat()`s since most things were in fact directories. Even if a failed open is actually slower, we're not going to get many of them and effectively halving the number of system calls is what really speeds things up here!